src/k8s-extension/azext_k8s_extension/partner_extensions/WorkloadIAM.py

```diff
@@ -0,0 +1,139 @@
+        local_authority = configuration_settings.pop(CONFIG_SETTINGS_USER_LOCAL_AUTHORITY, None)
+        join_token = configuration_settings.pop(CONFIG_SETTINGS_USER_JOIN_TOKEN, None)
+
+        # A trust domain name is always required
+        if trust_domain is None:
+            raise InvalidArgumentValueError(
+                f"Invalid configuration settings '{configuration_settings}'. Please provide a trust "
+                "domain name.")
+
+        if tenant_id is None:
+            raise InvalidArgumentValueError(
+                f"Invalid configuration settings '{configuration_settings}'. Please provide a "
+                "tenant ID.")
+
+        # If the user hasn't provided a join token, create one
+        if join_token is None:
+            if local_authority is None:
+                raise InvalidArgumentValueError(
+                    f"Invalid configuration settings '{configuration_settings}'. Either a join "
+                    "token or a local authority name must be provided.")
+            join_token = self.get_join_token(trust_domain, local_authority)
+        else:
+            logger.info("Join token is provided")
+
+        # Save configuration setting values to overwrite values in the Helm chart
+        configuration_settings[CONFIG_SETTINGS_HELM_TRUST_DOMAIN] = trust_domain
+        configuration_settings[CONFIG_SETTINGS_HELM_TENANT_ID] = tenant_id
+        configuration_settings[CONFIG_SETTINGS_HELM_JOIN_TOKEN] = join_token
+
+        logger.debug("Configuration settings value for Helm: %s", str(configuration_settings))
+
+        create_identity = True
+        extension = Extension(
+            extension_type=extension_type,
+            auto_upgrade_minor_version=auto_upgrade_minor_version,
+            release_train=release_train,
+            version=version,
+            scope=ext_scope,
+            configuration_settings=configuration_settings,
+            configuration_protected_settings=configuration_protected_settings
+        )
+        return extension, name, create_identity
+
+
+    def get_join_token(self, trust_domain, local_authority):
+        """
+        Invoke the az command to obtain a join token.
+        """
+
+        logger.info("Getting a join token from the control plane")
+
+        # Invoke az workload-iam command to obtain the join token
+        cmd = [
+            "az", "workload-iam", "local-authority", "attestation-method", "create",
+            "--td", trust_domain,
+            "--la", local_authority,
+            "--type", "joinTokenAttestationMethod",
+            "--query", "singleUseToken",
+            "--dn", "myJoinToken",
+            ]
+        cmd_str = " ".join(cmd)
+
+        try:
+            # Note: We can't use get_default_cli() here because its invoke() method
+            # always prints the console output, which we want to avoid.
+            result = subprocess.run(cmd, capture_output=True, shell=True)
```

Comment:
>subprocess.run(cmd, capture_output=True, shell=True) [](http://example.com/codeflow?start=21&length=52)

Can you pl. verify if this will work when the user is not a local admin?

Reply:
Do you mean in case the AZ CLI has been installed by someone else other than the user?

Reply:
I meant the CLI being invoked by a user who is non-admin.  Will they have permission to invoke a sub-process and create a token?

Reply:
Got it.  Thanks for sharing the details, Antonio1